### DISCUSSION TITLE
Surrogate model in IDAES

### DISCUSSION THREAD

**Discussion by abhi24x7** [IDAES/idaes-pse#1386] [General]
I have been familiarising myself with the framework now, and documentation is very elaborate and descriptive. Thank you IDAES community.

Using the IDAES framework, I now have modelled a packed bed reactor using customized non-Arrhenius rate equations, now I aim to model the same reactor using a ANN based surrogate model, embed it into the flowsheet inplace of the PFR block and run optimization over the flowsheet.
For getting started with ANN based surrogate models with IDAES, I have referred 
https://idaes.github.io/examples-pse/latest/Examples/SurrMod/FlowsheetOptimization/Keras_flowsheet_optimization_doc.html

However this is just illustrating the ANN based model for a single unit, i.e. reformer. It doesn't have any other unit operations ( and related physical property package) in the flowsheet. Please correct me if I am wrong. 

I was wondering if embedding a ANN based model for the Packed bed reactor is possible in IDAES (assuming it has to be linked with the physical property package in someway), any support would be highly appreciated. Please let me know if further clarity is needed.

Thank you very much in advance!

**Comment by andrewlee94** [comment]
@abhi24x7 It is definitively possible to embed an ANN based model into a flowsheet, but there are a few ways to do it. Based on your comment above, it looks like you are thinking about creating an ANN for the entire PFR with reactions (which is a good idea if you have complex reactions).

Firstly, your ANN based model does not have to be linked to any specific property package. IDAES supports having different units in the process use different property packages, or even no property package at all if it makes sense. The one thing your surrogated-based model will have to do is declare `Ports` which will be used to connect to the other units in the flowsheet.

Next, when it comes to connecting the `Ports` to other units in the flowsheet, the `Ports` need to have the same contents; basically they have to have the same set of variables and indices and consistent references states. Using the same property package in each unit model normally ensures this, but it is also possible to "translate" between different property packages (or `Port` definitions).

So, in your case you have two options.

1. Define the `Ports` in your ANN based model to be consistent with the property package you are using up- and down-stream. This is probably easy in your case, as you will be building the ANN from a PFR using the same property package.
2. Alternatively, IDAES also has a `TranslatorBlock` which allows you map the `Ports` in one property package to those in another. Basically, this is just an empty unit model with an inlet Port defined using one property package and the outlet using the other, and it is then up to you to write a set of constraints which do the mapping (hint: if you do this, always equate temperature and not enthalpy as temperature is absolute whereas enthalpy is based on a reference state).

**Reply by abhi24x7** [reply]
Dear @andrewlee94,
Thank you very much for your reply and assurance that ANN based PFR is possible, could you please tell me if there is any resource/example (apart from the AutoThermalReformer surrogate model), that shows how to embed a ANN based PFR inplace of PFR block in a flowsheet?

Secondly, I am planning to make the ANN using Pytorch, is it possible to use a pytorch model in IDAES or do I have to make a keras model like the one describe in the AutoThermalReformer example?

**Reply by andrewlee94** [reply]
@abhi24x7 Here is an example of using the IDAES `SkeletonUnitModel` to implement a surrogate model into a flowsheet: https://idaes-examples.readthedocs.io/en/latest/docs/unit_models/operations/skeleton_unit_doc.html

At the moment, I think we only support Keras (although I am not sure as I do not help maintain those tools).

**Reply by abhi24x7** [reply]
Hello,
Thank you for all your previous replies!

With regards, to my ANN based PFR model. I am slightly stuck in figuring out a method to embed it into IDAES, for better explanation from your end, I would like to explain you about the ANN model:

`Inputs` : length of reactor (z),  inlet Temp, inlet Pressure,  inlet feed flowrate of CH4,  inlet feed flowrate of O2,  inlet feed flowrate of CO2,  inlet feed flowrate of of H2O.

`Outputs` : outlet Temp, outlet Pressure,  inlet feed flowrate of CH4,  inlet feed flowrate of O2,  inlet feed flowrate of CO2,  inlet feed flowrate of H2O

Since the variable "z" is not a state variable that can be transferred from one unit to another, how do I deal with this issue?

While going through the IDAES documentation I came to know that, ANN type surrogate models can be embedded using OMLT (as shown in the ATR example).  But there is only one unit and no other unit operations. The reason I am stating this is because, I am planning to include other units like compressors that use a property package with state variables as: `FTPx` 
How can I make the inputs and outputs of the ANN model consistent with the property package that is going to be used in other unit blocks?

Thank you and looking for your reply!
Thank

**Reply by andrewlee94** [reply]
@abhi24x7 When you embed your ANN into IDAES, the want to map all your ANN inputs and outputs to Pyomo `Vars`. You can then decide how to expose all of these to the user; the feed and outlet stream will be mapped to `Ports` so that they can be connected up and down stream, but you can have other variables (such as `z`) that are not included in the `Ports` and are treated as `Vars` the user is expected to fix themselves (e.g. `volume` in the CSTR unit model: https://idaes-pse.readthedocs.io/en/stable/reference_guides/model_libraries/generic/unit_models/cstr.html?highlight=CSTR).

**Comment by abhi24x7** [comment]
@andrewlee94,
Thank you very much for your prompt reply and suggestion. 

First and foremost, I apologize if I have repeated certain things from the previous replies, its just that I want to understand the development of the model in a better manner and I have been facing slight difficulty in doing so.

Having said that, I plan to use `OMLT` framework to include my trained ANN into Pyomo formulation, and the ANN is trained to accommodate the inputs and outputs in the following order:

`Inputs` : length of reactor (z), inlet Temp, inlet Pressure, inlet feed flowrate of CH4, inlet feed flowrate of O2, inlet feed flowrate of CO2, inlet feed flowrate of of H2O.

`Outputs` : outlet Temp, outlet Pressure, outlet feed flowrate of CH4, outlet feed flowrate of O2, outlet feed flowrate of CO2, outlet feed flowrate of H2O.

So based on your reply, when you say , "_map all your ANN inputs and outputs to Pyomo_ `Vars`", does it mean that I make a different python script `ANN_PFR.py` (different from where I have defined the rest of the flowsheet) where in load the trained ANN model as an OMLT formulation?
Or do I load it in the Jupyter notebook where the rest of the flowsheet is established, like how it is shown in [this](https://idaes-examples.readthedocs.io/en/2.1.0/docs/surrogates/omlt/keras_flowsheet_optimization_doc.html)
![image](https://github.com/IDAES/idaes-pse/assets/66618073/2e135665-9232-4353-b499-f011210ac1b7)

Secondly, I had a look at the CSTR unit model [source code](https://idaes-pse.readthedocs.io/en/stable/_modules/idaes/models/unit_models/cstr.html#CSTRData) and source codes of other unit models as well, I see that all of them have a `CONFIG = UnitModelBlockData.CONFIG()` which stores Balance equations (material, energy & momentum) which are called in the `build()` method on the `ControlVolume0D` which makes sense.

However in my case do I need to implement the `CONFIG = UnitModelBlockData.CONFIG()` as well? Because I don't see why I should implement the balance equations, please correct me if I am wrong. 

Another thing is that in the rest of the flowsheet, the state variables that are being passed are `FTPx`, I cannot understand how can I directly link the ANN model to the other units in the flowsheet as these are not the variables "directly" being computed by the ANN, so I wonder how do I pass the `FTPx` state variables.
One _indirect_ way I thought about this is to write a custom function that takes in the FTPx values from the previous unit, converts (as well as scales it) to the type of input that the ANN requires and once predicted, another custom python function would unscale the predicted values and compute the output FTPx as required by the rest of the flowsheet. 
I feel that there might be a more smarter way to think about this, because based on the current description, I might run into issues/errors beyond my understanding.

In the CSTR unit model source code I see that the ports are added by `self.add_inlet_port()` and `self.add_outlet_port()` and I presume that these ports read the inlet and outlet state variables as defined by the CONFIG. I also wanted to know how can I pass the variables computed by the ANN that the ports can easily transfer to the other units of the flowsheet.

Thank you for reading through my issue, and apologies for the long text I just tried to describe it in the best possible manner I can.

Looking forward to your reply.

**Reply by andrewlee94** [reply]
@abhi24x7 Before I try to answer some of your specific questions, one important thing to realize is that there is a separation between the Python code that builds the model and the actual model objects that are constructed. I.e., the code you use to build the model matters little as long as the underlying model structure is the same. Thus, details of implementation are ultimately up to what works best for you.

I am not sure if you have seen it yet, but this is the best example of what you are trying to do: https://github.com/IDAES/examples/blob/main/idaes_examples/notebooks/docs/unit_models/operations/skeleton_unit.ipynb

1. Exactly where and how you run the code doesn't matter much in the end. You can do it all in one script, in the notebook, or break it into smaller sub modules if you want. Being this is Python, there are many ways you can achieve what you are trying to do, and whilst some are better than others if it works in the end it is OK.
2. Second to this, IDAES tries to use a duck-typing type approach wherever possible. For unit models, this means that in order to be "flowsheet-compatible" there area few outward facing things that must be present, but what happens inside the unit-model `Block` matters less. I.e. you ideally want your new custom block to have `Ports` so that it can connect to other unit models (but there are ways around even this), and a few outward facing design variables the user needs to provide, but the internals of the block (the ANN and how it is linked) is up to you.
3. The key factor in linking an external surrogate modeling into an IDAES flowsheet is recognizing that the flowsheet represents everything as `Vars`, and that the ANN in this case is basically a set of black-box constraints that link the `Vars` representing the inputs to those representing the outputs. These `Vars` can represent different things in the flowsheet, and can be used in different ways; e.g. some can be used in `Ports `representing flows in and out of a specific unit operations, whilst others might represent characteristics of a given unit operation.
4. Looking at the code for the existing unit models is not likely to help you much in this. As alluded to in 1), this is just the Python code that builds the model objects, and is just a standard format we use to do this. Things like the `CONFIG` block mean far less in your case where you have a one-off custom, black-box model. What is important however is the resulting model structure - note the CSTR has two `Ports` (inlet and outlet) which contain the "state variables" for the streams, and there are additional unit-model level design variables the user must fix (e.g. volume).
5. Another important thing to note here is that the model is not solved sequentially but simultaneously. This means that you cannot embed Python code into your actual model easily (this is why OMLT exists). So, your idea about having code that maps IDAES Vars to the ANN and then maps results back out will not work as the solver will not see any of that (your Python code would be run once with the initial state and that would be it).
6. Once you link the `Vars` to your ANN via OMLT, they will be directly linked. OMLTs role is to perform that linking for you so that when you call a solver, it treats your ANN as a set of additional constraints that link the inlet `Vars` to the outlet `Vars.
7. The contents of the `Ports` are not defined by the `CONFIG` block, but by the property package you are using. The `add_inlet_port` and `add_outlet_port` methods will probably not be of any use to you directly as you do not have a property package in the custom unit model. However, if you follow the code you will see how you can add `Vars` to a `Port`, which is what you will want to do; you will create a `Port` for the inlet, and then add all the associated `Vars` to it (and the same for the outlet). One important note here; you need the contents of the `Port` to match that of what it will be linking to, so you need to match this to the property package you are using up and downstream (including the names you assign).

**Reply by abhi24x7** [reply]
Dear @andrewlee94,

Thank you very much for your elaborative previous reply. Now I have fair idea about how the `Ports` and 
I possibly think that my approach would be to use a `SkeletonUnitModel` and 'OmltBlock'.
Is it possible to include an `OmltBlock` inside a `SkeletonUnitModel`? Is there any example in the IDAES repository that does a near similar kind of thing?

**Reply by andrewlee94** [reply]
Unfortunately I don't think we have an example yet for including `OMLT` in a `SkeletonUnitModel`. however it is definitely possible to put an `OmltBlock` inside a `SkeletonUnitModel` (generally Blocks can be freely nested).

@bpaul4 or @rundxdi might have some examples of where they have used OMLT in process flowsheets.

**Reply by bpaul4** [reply]
Hello @abhi24x7, we have an example of integrating OMLT models via the `KerasSurrogate` in an IDAES flowsheet: https://github.com/IDAES/examples/blob/main/idaes_examples/notebooks/docs/surrogates/omlt/keras_flowsheet_optimization.ipynb. This example demonstrates a surrogate that replaces a Gibbs reactor for an autothermal reformer, and you may wish to do something similar for your application.  If you'd like to use a `SkeletonModel` for more flexibility, we have an example utilizing a hard-coded surrogate model (not a neural network): https://github.com/IDAES/examples/blob/main/idaes_examples/notebooks/docs/unit_models/operations/skeleton_unit.ipynb.

**Reply by abhi24x7** [reply]
Dear @andrewlee94,

Thank you very much for your previous reply and suggestions. I have kickstarted the development of the ann embedded model by referring to `SkeletonUnitModel` example as in [here](https://github.com/IDAES/examples/blob/main/idaes_examples/notebooks/docs/unit_models/operations/skeleton_unit.ipynb).

By following the procedure as given in the link. I developed the following code block [surrogate_model_v1.txt.txt](https://github.com/IDAES/idaes-pse/files/15143032/surrogate_model_v1.txt.txt)

I fairly understood the concept of ports here and how they transfer the state variables from one unit to another. 

I get the following error when I try to run this block of code: 
`ERROR: Rule failed when initializing variable for Var fs.sm.mole_frac_comp_in
with index carbon_monoxide: KeyError: "Index 'carbon_monoxide' is not valid
for indexed component 'fs.sm.mole_frac_comp_in'"
ERROR: Constructing component 'fs.sm.mole_frac_comp_in' from data=None failed:
        KeyError: "Index 'carbon_monoxide' is not valid for indexed component
        'fs.sm.mole_frac_comp_in'"`

Could you please make me understand what seems to be the issue here?

Secondly, in the `SkeletonUnitModel`, when I go to the property package used over there `eg_h2o_ideal.py` and scroll towards the bottom I see that the state variables that they pass are `FpcTP` and the same variables are passed through the ports (which is also what you meant in the replies earlier). However, I want to understand what is exactly being done in these kind of lines of code:
`m.fs.pervap.flow_in = Var(
    m.fs.time,
    m.fs.pervap.phase_list,
    m.fs.pervap.comp_list,
    initialize=1.0,
    units=pyunits.mol / pyunits.s,
)`

I am asking you this, is because in my case the state variables that are passed in my Property package are `FTPx` and I have tried to implement the same as in the file attached, however I fail to completely understand whether I have correctly implemented the variables like `mole_frac_comp_in`, `mole_frac_comp_out`, `flow_in` and `flow_out` or not. Could you please suggest on what seems to be done incorrectly?
With regards to the `phase_list` and `comp_list` they are already there in my property package, why do I define them again when I declare the variables? 

Looking forward to your reply!

**Reply by andrewlee94** [reply]
1. From that error, it appears that you have a mismatch between the indexing set for `.mole_frac_comp_in` and the initial values you are providing it. Notably, it looks like you have provided an initial value for `carbon_monoxide`, but Pyomo is telling you that is not in the indexing set. Start by checking the indexing set to make sure `carbon_monoxide` is there and is spelt correctly.
2. That line of code is creating a new Pyomo `Var`, indexed by the flowsheet time domain (`m.fs.time`), the phase list from the associated property package (`m.fs.pervap.phase_list`), and component list from the property package (`m.fs.pervap.comp_list`).
3. Without seeing your code, it is hard for me to say if you've done everything correctly or not, and even then it might be hard to interpret everything your are doing. The place to start is by using the Pyomo `display()` method to see if those variables have been added to your
 model, and if they have the correct indexing sets and values.
4. You are not declaring the indexing sets again when you create the variable, you are telling Pyomo that these sets are to be used to index this variable. Note that you are passing the existing instances of these sets to the `Var()` class, so they already exist.

**Reply by abhi24x7** [reply]
Dear @andrewlee94,

Thank you for your quick reply! I understand all your reasoning, with respect to point 3, is there any section of code you want to specifically look at? I have also attached the code block in the previous reply as well named "Surrogate_model_v1.txt.txt"

**Reply by andrewlee94** [reply]
Ah - I see the problem. The issue is that when you initialize the Vars, the dict you need to provide needs to match the full indexing set for the Var. I.e., `carbon_monoxide` is not a valid key as `m.fs.sm.mole_frac_comp_in` is indexed by time, phase and component. What you need is:

```
m.fs.sm.mole_frac_comp_in = Var(
    m.fs.time,
    m.fs.sm.phase_list,
    m.fs.sm.comp_list,
    initialize = {
       (0, "Vap", 'carbon_monoxide'): 1e-2,
       (0, "Vap", 'carbon_dioxide'): 1e-2,
       (0, "Vap", 'hydrogen'): 1e-2,
       (0, "Vap",  'methane'): 1e-2,
       (0, "Vap", 'methanol'): 1e-2,
       (0, "Vap",  'water'): 1e-2,
       (0, "Vap", 'nitrogen'): 1e-2},
    units = pyunits.dimensionless)
```

One other thing to note on naming and conventions in IDAES: the name `mole_frac_comp_in` indicates that this is mole fractions indexed by time and component list, however you also included the phase list in the indexing sets. I am not sure which you actually want (this will depend on the up- and down-stream property packages, but if you need to include the phase as an index then the standard name would be `mole_frac_phase_comp`.

**Reply by abhi24x7** [reply]
Dear @andrewlee94,

Thank you very much for your previous reply.
I was also looking at other different ways to embed the surrogate model in IDAES, as I was not gaining much success with `SkeletonUnitModel`. I came across the source code of the scientific article titled _Process Flowsheet Optimization with Surrogate and Implicit Formulations of a Gibbs Reactor_ 

Link to article: [https://arxiv.org/html/2310.09307v2](https://arxiv.org/html/2310.09307v2)
Link to source code: [(https://github.com/Robbybp/surrogate-vs-implicit/tree/main)](https://github.com/Robbybp/surrogate-vs-implicit/tree/main)

In this work, they also use a Neural network trained in Keras and they embed it into IDAES. This something very similar to what I aim to do  in my case, with the difference being, here they use a Gibbs Reactor as the underlying reactor model and I use the PFR. Even they are using the state variables `FTPx`.

When I look at the way they have implement the "embedding" of the ANN based model (link to the [code](https://github.com/Robbybp/surrogate-vs-implicit/blob/main/svi/auto_thermal_reformer/nn_flowsheet.py)), I found something interesting, firstly, they haven't used any `SkeletonUnitModel`, 
- They have declared a `SurrogateBlock`, 
- declared the inputs and outputs variables (as a python list)
- And finally they have loaded the trained ANN model and have build the formulation.

Before they have done the above said steps they have done this:
![image](https://github.com/IDAES/idaes-pse/assets/66618073/b5d5bc67-4be9-408e-8880-e6472f9b7bd7)

From what I understood when I looked up, what this could do is that, the above lines "deactivates" all the constraints related to the original Gibbs reactor. Is this right? Additionally, what I also assume that once the constraints are deactivated, it essentially "by-passes" the original Gibbs reactor calculation routine, which is critical to shift it to the predictions by the Surrogate model.

Secondly, I noticed that the outputs of the ANN are directly **referenced** from the variables of the original IDAES Gibbs Reactor, like as shown:
![image](https://github.com/IDAES/idaes-pse/assets/66618073/47085f21-300d-4c3e-8f82-c7c0ee331a7c)
What is the need to do the same?

Now, coming to my case
(Simulating just the ANN using OMLT, no integration with IDAES)
I am using an onnx file of a neural network (about which I elaborately describe in my previous replies). I then use a `OmltBlock` to make it to a formulation. And I then connect the inputs that are already in the form which the neural network needs, I solve it using a solver and it readily predicts the outputs!

(Integration of ANN with IDAES)
To keep it simple, I just have a  `Feed`, `PFR` and `Product` block, I have declared all the Degrees of Freedom for all the unit blocks, connected the arcs and have Initialized it. (Things are good till here!)

Now to implement the ANN in the IDAES, I tried to implement a similar approach to what is done in the Gibbs ATR example in my case, where I deactivate all the constraints related to my `PFR` and then declare the respective output variables, as shown


![image](https://github.com/IDAES/idaes-pse/assets/66618073/9e96d9fc-8328-4cb3-b069-a24303833041)

After this, I declare the scaling factors (to scale the inputs for the ANN and to unscale outputs from the ANN) as Pyomo `Param`.  I then connect the scaled variables to the inputs of the ANN OmltBlock
Kindly note that: my ANN takes length of reactor (z), T, P and molar flowrates of the component (I calculate this by multiplying flowrate by mole fraction and pass it to NN)
![image](https://github.com/IDAES/idaes-pse/assets/66618073/8d2dcd05-3d9c-4ca1-8a25-2bfa433f9eb1)

Similarly, I connect the scaled outputs of the Neural Network, to the scaling factors to unscale it and pass it to the state vars of R101
![image](https://github.com/IDAES/idaes-pse/assets/66618073/470c5683-1951-4d7b-8091-332ebd295d63)

After this when I attempt to solve the flowsheet, I get the following error:
![image](https://github.com/IDAES/idaes-pse/assets/66618073/1aa57d3f-35ca-4de8-92be-3c970aad63e3)

When I looked into the model using `m.fs.nn.display()`, I see that 
![image](https://github.com/IDAES/idaes-pse/assets/66618073/e43f330e-ed23-41e9-a433-cc0e2204061b)

The inputs have been scaled properly and fed as the input to the ANN, however the outputs are simply 0 (which somewhat related to the error the solver is throwing.)

This is the problem I am facing, I tried to elaborate it as much as possible and I have been trying to debug but couldn't reach any further. Could you please suggest me what can be done here?

Thank you and looking forward to your reply!!

**Reply by andrewlee94** [reply]
So, first of all if the error you are getting is a solver failure, and notably an infeasible solution. This means that the code built the model successfully, but that it was unable to solve which means the issue lies in the formulation of the model and not in the syntax of the code.

The first place to start with problems like these is the Diagnostics Toolbox: 

Documentation: https://idaes-pse.readthedocs.io/en/stable/explanations/model_diagnostics/index.html
Tutorial/Example: https://idaes-examples.readthedocs.io/en/latest/docs/diagnostics/diagnostics_toolbox_doc.html

These tools will help you look at your model and try to work out where the problem lies. Looking at the error message you are getting, a few things jump out:

1. It failed in zero iterations.
2. It says that there is an infeasibility in the triangular part of the problem. This means you have a system of equations in the model that can be solved sequentially (e.g. A = B, B=C, etc.) which is infeasible - this is probably not related to the ANN itself but the other equations in the model.

You should start by first making sure the degrees of freedom in the model are 0 (non-zero degrees of freedom can mask underlying issues) and then check the bounds on all your variables. It is likely you are trying to set a value outside a bound and this is causing the infeasibility. The Diagnostics Toolbox can help you identify these,
